### PR TITLE
[feature/extended-share-sheet-reorg] Share Sheet Extension (re-organized)

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Review / Reorg Note

This PR is a code-reviewed & bugfixed version of `feature/extended-share-sheet` built on top of `feature/appshared-reorg` (#751, which in turn is factoring out the UI and infrastructure code necessary for use in extensions from `ownCloud.app`  to `ownCloudAppShared.framework`).

The code review notes and relevant code changes can be found in b454ce65bc837646ea5e5a21f7b62cfcc21f9ecc.

Below is a reproduction of the original `feature/extended-share-sheet` PR's text:

---

## Description

Implementation for the share sheet extension. 
With this extension it is possible to import more than one file and it is available in all apps.

For this implementation I had to change  a lot of internal structure. UI code and needed files was moved from the app target to the shared framework target.

## Related Issue
#539 #650

## Motivation and Context
The current import implementation can only import one file and is not available in all share sheet apps like Photos.app. This replacement solves this problems and the import is possible without switching to the app.

## How Has This Been Tested
- Open Photos.app and select one or more items
- Tap on `Share` button
- Select `ownCloud` app

### Additional tests

- Open Safari.app
- Open a website
- Tap on `Share` button
- Choose `Options` and select `PDF`
- Select `ownCloud` app

- Add a Passcode to the app
- Remove all accounts
- Add a new account
- Test the whole app, because a lot of code was moved

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA

Test Plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.4.0/ShareSheet%20II%20.md

BUGS & IMPROVEMENTS

- [x] (1) Some files are lost https://github.com/owncloud/ios-app/pull/650#issuecomment-659388302
- [X] (2) Sharing text https://github.com/owncloud/ios-app/pull/752#issuecomment-666236806 [FIXED]
- [x] (3) Provider name in Files App https://github.com/owncloud/ios-app/pull/752#issuecomment-666241134 [FIXED]
- [X] (4) Options repeated in share menu https://github.com/owncloud/ios-app/pull/752#issuecomment-666273459 [WONT FIX]

### CODE REVIEW

File: ownCloud Share Extension/ShareViewController.swift [hunk 1]
```diff
@@ -66,36 +73,44 @@ class ShareViewController: MoreStaticTableViewController {
...
 			if bookmarks.count > 1 {
-				let rowDescription = StaticTableViewRow(label: "Choose an account and folder to import the file into.".localized, alignment: .center)
+				let rowDescription = StaticTableViewRow(label: "Choose an account and folder to import into.".localized, alignment: .center)
 				actionsRows.append(rowDescription)
 
 				for (bookmark) in bookmarks {
 					let row = StaticTableViewRow(buttonWithAction: { (_ row, _ sender) in
 						self.openDirectoryPicker(for: bookmark, pushViewController: true)
-					}, title: bookmark.shortName, style: .plain, image: Theme.shared.image(for: "owncloud-logo", size: CGSize(width: 25, height: 25)), imageWidth: 25, alignment: .left)
+					}, title: bookmark.shortName, style: .plain, /* make brandable before re-adding --> image: Theme.shared.image(for: "owncloud-logo", size: CGSize(width: 25, height: 25)), imageWidth: 25, */ alignment: .left)
```
Comment (hosy) [inline ownCloud Share Extension/ShareViewController.swift:82]: I will change this directly and using the brandable icon `bookmark-icon.png` instead of.

**Comment by jesmrec** [comment]
### (2)

1. Open Safari and any web site
2. Select some text in the content
3. Select `Share` and `ownCloud`
4. Select account and target folder

Current: text not formatted
Expected: text formatted in any way

iPadAir v13, iPhoneXR v13

`c1a8fad2`

**Comment by jesmrec** [comment]
### (3)

I think that this one was already repoted by you. I add here, to track the fix:

Current: In Files app, the ownCloud accounts is displayed as `owncloud File Provider` 
Expected: In Files app, the ownCloud accounts is displayed as `owncloud` (or the app name) 

commit `c1a8fad2`

**Comment by jesmrec** [comment]
Regarding (1)., i moved a step further and tried to share a higher number of files (~300) with same bad result (only 179 finished). The app is closed, so, i thought that some background issues can happen due to the known problem in iOS with NSURLSession. Is that right? does it make sense to test with such huge number of files, with not enough time to upload everything before app goes to background?

**Comment by jesmrec** [comment]
### (4)

As well as (3), reported by you, just for me not to forget to check the fix:

Current: in Share menu, owncloud appears twice: `ownCloud` and  `Copy to ownCloud`
Expected: in Share menu, owncloud appears once: `ownCloud`

I see the problem only in iPhone. Not reproducible in iPad

commit `c1a8fad2`

**Comment by hosy** [comment]
@jesmrec 
Here are my comments to your findings (commit: 96be76a):
(1) @felix-schwarz did some changes, please test this again
(2) selected text will now be saved als `txt` file. Formatting cannot be preserved. I also added the possibility to save images, links from a website.
(3) The display name of the File Provider extension was changed to `ownCloud`
(4) I didn´t find a way to remove the duplicated entry. The problem is, when we remove handling document types, the app cannot receive files via AirDrop. The legacy iOS app has two entries too.

**Comment by jesmrec** [comment]
My last review  with commit `961e9784` and sdk `1a93f7e`:

> (1) @felix-schwarz did some changes, please test this again

I gave a new try with 100 files. After sharing with oC, and opening the app, most of them are stucked:

Aug-03-2020 14-58-47

after killing the app and reopening again, uploads start correctly, but only 51 are finished when `Status` gets empty. These are the logs i caught:

[ownCloud_3_Aug_2020_at_14_55_49.log.txt.zip](https://github.com/owncloud/ios-app/files/5016083/ownCloud_3_Aug_2020_at_14_55_49.log.txt.zip)

i hope it helps.

> (2) selected text will now be saved als txt file. Formatting cannot be preserved. I also added the possibility to save images, links from a website.

Selecting text and sharing with oC app leads to an infinite "Saving":

Aug-03-2020 14-50-32

Saving links and images from a web site creates a `import.public.url` file. Is that correct? is this the expected one?

> (3) The display name of the File Provider extension was changed to ownCloud

 i still see:

Screenshot 2020-08-03 at 13 06 47

> (4) I didn´t find a way to remove the duplicated entry. The problem is, when we remove handling document types, the app cannot receive files via AirDrop. The legacy iOS app has two entries too.

We'll let it be

**Comment by hosy** [comment]
@jesmrec please can you re-check with a clean install. For me everything is working fine with commit 96be76a8

**Comment by jesmrec** [comment]
Problem was on my side, checking the wrong branch `feature/extended-share-sheet` . Will review with the current one

**Comment by jesmrec** [comment]
app commit: `ce0167d`
sdk commit: `1e2e353`

> (1) @felix-schwarz did some changes, please test this again

50 files ✅ 
100 files ✅ 

With a higher number of files, the behaviour changes from one try to other one. F. ex, with 150 files, some of them were missed, but in a second try everything was uploaded... not deterministic. I caught some logs (200 shared, but only 122 as final result):

[ownCloud_4_Aug_2020_at_12_58_53.log.txt.zip](https://github.com/owncloud/ios-app/files/5021729/ownCloud_4_Aug_2020_at_12_58_53.log.txt.zip)

Status view was totally cleaned.

> (2) selected text will now be saved als txt file. Formatting cannot be preserved. I also added the possibility to save images, links from a website.

Text, images and links correctly saved. As improvement for the future, let the user decide the name of the txt file. Will it be feasible (in another issue, for sure)?

> (3) The display name of the File Provider extension was changed to ownCloud

fixed

> (4) I didn´t find a way to remove the duplicated entry. The problem is, when we remove handling document types, the app cannot receive files via AirDrop. The legacy iOS app has two entries too.

As discussed in the call, the entry `Copy to owncloud` will remain ftm.

**Comment by hosy** [comment]
@jesmrec please can you share your files, which you tried to upload?

**Comment by jesmrec** [comment]
I shared files from Gallery app or Files App, 150 pics/videos (same files in both cases) that together takes more than 500MB. Should i fetch them to send you? it should not be different with files in your Gallery.

**Comment by jesmrec** [comment]
Latest checks:

- [X] Importing progress bar 
- [X] Error handling
- [X] In case a huge number of files is shared, everything is correctly imported (tested with different number of files till ~500 imports from Photos)

From my side, it is OK to move forward this iteration.

**Pull request merged**